generate/tp.py

```diff
@@ -0,0 +1,222 @@
+    tensor_parallel_linear(fabric, attn.proj, "rowwise")
+    attn.register_forward_hook(partial(all_reduce_output, fabric.world_size))
+
+
+def all_reduce_output(world_size: int, module: torch.nn.Module, ins, outs) -> torch.Tensor:
+    return all_reduce(outs, "sum", list(range(world_size)))
+
+
+def tensor_parallel(fabric: L.Fabric, model: GPT) -> GPT:
+    for block in model.transformer.h:
+        tensor_parallel_mlp(fabric, block.mlp)
+        tensor_parallel_attn(fabric, block.attn)
+
+    # update the config values to the shard sizes
+    # this is only relevant for `tensor_parallel_attn`, but it needs to run only once
+    world_size = fabric.world_size
+    model.config.n_head //= world_size
+    model.config.n_embd //= world_size
+    model.config.n_query_groups //= world_size
+
+    return model
+
+
+@torch.inference_mode()
+def main(
+    prompt: str = "What food do llamas eat?",
+    *,
+    num_samples: int = 1,
+    max_new_tokens: int = 50,
+    top_k: Optional[int] = 200,
+    temperature: float = 0.8,
+    checkpoint_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b"),
+    quantize: Optional[Literal["bnb.nf4", "bnb.nf4-dq", "bnb.fp4", "bnb.fp4-dq"]] = None,
+    precision: Optional[str] = None,
+    compile: bool = False,
+) -> None:
+    """Generates text samples based on a pre-trained model and tokenizer.
+
+    Args:
+        prompt: The prompt string to use for generating the samples.
+        num_samples: The number of text samples to generate.
+        max_new_tokens: The number of generation steps to take.
+        top_k: The number of top most probable tokens to consider in the sampling process.
+        temperature: A value controlling the randomness of the sampling process. Higher values result in more random
+            samples.
+        checkpoint_dir: The checkpoint directory to load.
+        quantize: Whether to quantize the model and using which method:
+            - bnb.nf4, bnb.nf4-dq, bnb.fp4, bnb.fp4-dq: 4-bit quantization from bitsandbytes
+            for more details, see https://github.com/Lightning-AI/lit-gpt/blob/main/tutorials/quantize.md
+        precision: Indicates the Fabric precision setting to use.
+        compile: Whether to compile the model.
+    """
+    precision = precision or get_default_supported_precision(training=False)
+
+    plugins = None
+    if quantize is not None:
+        if compile:
+            raise NotImplementedError  # untested
+        if "mixed" in precision:
+            raise ValueError("Quantization and mixed precision is not supported.")
+        dtype = {"16-true": torch.float16, "bf16-true": torch.bfloat16, "32-true": torch.float32}[precision]
+        plugins = BitsandbytesPrecision(quantize[4:], dtype)
+        precision = None
+
+    # set "ddp" as the strategy for the launching functionality, but there's no data-parallelism
+    fabric = L.Fabric(devices="auto", strategy="ddp", precision=precision, plugins=plugins)
+    fabric.launch()
```

Comment:
We should expose the fabric launcher or have a "nicer" way to use fabric just as a launcher.

The alternative to this hack is to instead use `torchrun`, but that is not lightning-y and is quite verbose